src/ShadyCSS.js

```diff
@@ -146,52 +149,74 @@ export let ShadyCSS = {
-        // update instance if native shadowdom
-        if (this.nativeShadow) {
-          let style = host.shadowRoot.querySelector('style');
-          style.textContent = StyleTransformer.elementStyles(host, styleInfo.styleRules);
+    let template = templateMap[is];
+    if (this._forceRecalc || styleInfo.shouldRecalculate || template._applyShimInvalid) {
+      if (this.nativeCss) {
+        if (template && template._style && template._applyShimInvalid) {
+          // update template
+          if (!template._invalidating) {
+            ApplyShim.transformRules(template._styleAst, is);
+            template._style.textContent = StyleTransformer.elementStyles(host, styleInfo.styleRules);
+            StyleInfo.validate(is);
+          }
+          // update instance if native shadowdom
+          if (this.nativeShadow) {
+            let root = host.shadowRoot;
+            if (root) {
+              let style = root.querySelector('style');
+              style.textContent = StyleTransformer.elementStyles(host, styleInfo.styleRules);
+            }
+          }
+          styleInfo.styleRules = template._styleAst;
+        }
+        this._updateNativeProperties(host, styleInfo.overrideStyleProperties);
+      } else {
+        this._updateProperties(host, styleInfo);
+        if (styleInfo.ownStylePropertyNames && styleInfo.ownStylePropertyNames.length) {
+          this._applyStyleProperties(host, styleInfo);
         }
-        styleInfo.styleRules = template._styleAst;
-      }
-      this._updateNativeProperties(host, styleInfo.overrideStyleProperties);
-    } else {
-      this._updateProperties(host, styleInfo);
-      if (styleInfo.ownStylePropertyNames && styleInfo.ownStylePropertyNames.length) {
-        // TODO: use caching
-        this._applyStyleProperties(host, styleInfo);
       }
     }
-    if (hasApplied) {
-      let root = this._isRootOwner(host) ? host : host.shadowRoot;
-      // note: some elements may not have a root!
-      if (root) {
-        this._applyToDescendants(root);
-      }
+    styleInfo.shouldRecalculate = false;
+    let root = this._isRootOwner(host) ? host : host.shadowRoot;
+    // note: some elements may not have a root!
+    if (root) {
+      this._applyToDescendants(root.children);
     }
+    this._forceRecalc = false;
   },
-  _applyToDescendants(root) {
-    let c$ = root.children;
-    for (let i = 0, c; i < c$.length; i++) {
-      c = c$[i];
+  applyStyleInvalid(node) {
+    let styleInfo = StyleInfo.get(node);
+    if (styleInfo) {
+      styleInfo.shouldRecalculate = true;
+    }
+  },
+  _isElementNode(node) {
+    return node.nodeType === Node.ELEMENT_NODE;
+  },
+  _applyToDescendants(children) {
+    for (let i = 0, c; i < children.length; i++) {
+      c = children[i];
       if (c.shadowRoot) {
         this.applyStyle(c);
+      } else if (nativeCssVariables && c.localName === 'slot') {
```

Comment:
Let's consider not doing this as it's a perf hit (yes, that means the issue still exists).